File: atlas-perception-py/atlas_perception/imu/imu_sensor.py
import math
import time
from typing import Callable, Optional

from ..deterministic import seeded_range
from ..types import IMUData


class IMUSensor:
    def __init__(self):
        self._is_running = False
        self._data_callback: Optional[Callable[[IMUData], None]] = None
        self._sample_index = 0
        self._velocity = {"x": 0, "y": 0, "z": 0}
        self._orientation = {"roll": 0, "pitch": 0, "yaw": 0}

    async def start(self) -> None:
        self._is_running = True

    async def stop(self) -> None:
        self._is_running = False

    def capture_data(self) -> IMUData:
        seed = self._sample_index
        self._sample_index += 1
        t = self._sample_index * 0.01
        accel_x = math.sin(t * 1.3) * 0.3 + seeded_range(seed, -0.02, 0.02)
        accel_y = math.cos(t * 0.7) * 0.2 + seeded_range(seed + 1, -0.02, 0.02)
        accel_z = 9.81 + math.sin(t * 2.1) * 0.1 + seeded_range(seed + 2, -0.01, 0.01)
        self._velocity["x"] += accel_x * 0.01
        self._velocity["y"] += accel_y * 0.01
        gyro_x = math.sin(t * 1.3) * 0.003 + seeded_range(seed + 3, -0.001, 0.001)
        gyro_y = math.cos(t * 0.9) * 0.002 + seeded_range(seed + 4, -0.001, 0.001)
        gyro_z = math.sin(t * 0.5) * 0.004 + seeded_range(seed + 5, -0.001, 0.001)
        mag_x = math.cos(t * 0.3) * 30 + seeded_range(seed + 6, -1, 1)
        mag_y = math.sin(t * 0.3) * 30 + seeded_range(seed + 7, -1, 1)
        mag_z = 40 + math.sin(t * 0.2) * 5 + seeded_range(seed + 8, -1, 1)
        data = IMUData(
            acceleration={"x": accel_x, "y": accel_y, "z": accel_z},
            gyroscope={"x": gyro_x, "y": gyro_y, "z": gyro_z},
            magnetometer={"x": mag_x, "y": mag_y, "z": mag_z},
            timestamp=time.time(),
        )
        if self._data_callback:
            self._data_callback(data)
        return data

    def set_data_callback(self, callback: Callable[[IMUData], None]) -> None:
        self._data_callback = callback

File: atlas-perception_deprecated/IMU/IMUSensor.ts
import { seededRange } from "../../atlas-kernel/utils/deterministic";

export interface IMUData {
  acceleration: { x: number; y: number; z: number };
  gyroscope: { x: number; y: number; z: number };
  magnetometer?: { x: number; y: number; z: number };
  timestamp: number;
}

export class IMUSensor {
  private isRunning: boolean;
  private dataCallback?: (data: IMUData) => void;
  private sampleIndex = 0;
  private velocity = { x: 0, y: 0, z: 0 };
  private orientation = { roll: 0, pitch: 0, yaw: 0 };

  constructor() {
    this.isRunning = false;
  }

  async start(): Promise<void> {
    this.isRunning = true;
  }

  async stop(): Promise<void> {
    this.isRunning = false;
  }

  captureData(): IMUData {
    const seed = this.sampleIndex++;
    const t = this.sampleIndex * 0.01;

    const accelX = Math.sin(t * 1.3) * 0.3 + seededRange(seed, -0.02, 0.02);
    const accelY = Math.cos(t * 0.7) * 0.2 + seededRange(seed + 1, -0.02, 0.02);
    const accelZ = 9.81 + Math.sin(t * 2.1) * 0.1 + seededRange(seed + 2, -0.01, 0.01);

    this.velocity.x += accelX * 0.01;
    this.velocity.y += accelY * 0.01;

    this.orientation.roll += Math.sin(t * 1.1) * 0.002;
    this.orientation.pitch += Math.cos(t * 0.9) * 0.002;
    this.orientation.yaw += Math.sin(t * 0.5) * 0.003;

    const gyroX = Math.sin(t * 1.3) * 0.003 + seededRange(seed + 3, -0.001, 0.001);
    const gyroY = Math.cos(t * 0.9) * 0.002 + seededRange(seed + 4, -0.001, 0.001);
    const gyroZ = Math.sin(t * 0.5) * 0.004 + seededRange(seed + 5, -0.001, 0.001);

    const magX = Math.cos(t * 0.3) * 30 + seededRange(seed + 6, -1, 1);
    const magY = Math.sin(t * 0.3) * 30 + seededRange(seed + 7, -1, 1);
    const magZ = 40 + Math.sin(t * 0.2) * 5 + seededRange(seed + 8, -1, 1);

    const data: IMUData = {
      acceleration: { x: accelX, y: accelY, z: accelZ },
      gyroscope: { x: gyroX, y: gyroY, z: gyroZ },
      magnetometer: { x: magX, y: magY, z: magZ },
      timestamp: Date.now(),
    };
    this.dataCallback?.(data);
    return data;
  }

  setDataCallback(callback: (data: IMUData) => void): void {
    this.dataCallback = callback;
  }
}
